File: src/algorithms/dominatingSet.ts
import { GraphData, AlgorithmExecutionResult, AlgorithmStep } from '../types/graph';

// Greedy Minimum Dominating Set & Emergency Hub Location
export function runDominatingSetHubPlacement(
  graph: GraphData,
  coverageRadiusKm: number = 3.5
): AlgorithmExecutionResult {
  const startTime = performance.now();
  const steps: AlgorithmStep[] = [];
  const allNodeIds = Object.keys(graph.nodes);
  const coveredNodes = new Set<string>();
  const dominatingHubIds = new Set<string>();

  function getDistance(n1: { lat: number; lng: number }, n2: { lat: number; lng: number }): number {
    const dLat = (n2.lat - n1.lat) * 111.32;
    const dLng = (n2.lng - n1.lng) * 111.32 * Math.cos((n1.lat * Math.PI) / 180);
    return Math.sqrt(dLat * dLat + dLng * dLng);
  }

  // Precompute coverage neighborhood for each candidate node
  const neighborhood: Record<string, Set<string>> = {};
  for (const u of allNodeIds) {
    neighborhood[u] = new Set<string>([u]);
    const nodeU = graph.nodes[u];

    // Direct topological neighbors
    for (const edgeId of graph.adjacencyList[u] || []) {
      const edge = graph.edges[edgeId];
      if (!edge || edge.isBlocked) continue;
      const v = edge.source === u ? edge.target : edge.source;
      neighborhood[u].add(v);
    }

    // Spatial radius neighbors
    for (const v of allNodeIds) {
      if (u === v) continue;
      const nodeV = graph.nodes[v];
      if (getDistance(nodeU, nodeV) <= coverageRadiusKm) {
        neighborhood[u].add(v);
      }
    }
  }

  steps.push({
    stepIndex: 0,
    description: `Dominating Set optimizer initialized. Target: Determine minimum strategic relief hub locations covering all ${allNodeIds.length} vertices within $\\le ${coverageRadiusKm}$ km radius.`,
    highlightedNodeIds: [],
    highlightedEdgeIds: []
  });

  // Greedy Dominating Set: Pick node that covers the largest number of currently uncovered nodes
  while (coveredNodes.size < allNodeIds.length) {
    let bestCandidate: string | null = null;
    let maxNewlyCovered = 0;

    for (const candidate of allNodeIds) {
      // Prioritize existing shelters / hospitals
      const node = graph.nodes[candidate];
      const isPreferredFacility = node.type === 'shelter' || node.type === 'hospital' || node.type === 'depot';
      const bonus = isPreferredFacility ? 2 : 0;

      let newlyCoveredCount = 0;
      for (const neighbor of neighborhood[candidate]) {
        if (!coveredNodes.has(neighbor)) {
          newlyCoveredCount++;
        }
      }

      const score = newlyCoveredCount + bonus;
      if (score > maxNewlyCovered) {
        maxNewlyCovered = score;
        bestCandidate = candidate;
      }
    }

    if (!bestCandidate || maxNewlyCovered === 0) {
      // Pick any remaining uncovered node
      for (const u of allNodeIds) {
        if (!coveredNodes.has(u)) {
          dominatingHubIds.add(u);
          coveredNodes.add(u);
        }
      }
      break;
    }

    dominatingHubIds.add(bestCandidate);
    const newCoveredInStep: string[] = [];
    for (const neighbor of neighborhood[bestCandidate]) {
      if (!coveredNodes.has(neighbor)) {
        coveredNodes.add(neighbor);
        newCoveredInStep.push(neighbor);
      }
    }

    if (steps.length < 30) {
      steps.push({
        stepIndex: steps.length,
        description: `Selected "${graph.nodes[bestCandidate]?.name}" as Emergency Dominating Hub #${dominatingHubIds.size}. Covered ${newCoveredInStep.length} new neighborhood zones.`,
        highlightedNodeIds: [bestCandidate, ...newCoveredInStep],
        highlightedEdgeIds: []
      });
    }
  }

  const nodeAttrs: Record<string, { isDominatingHub: boolean }> = {};
  for (const id of allNodeIds) {
    nodeAttrs[id] = { isDominatingHub: dominatingHubIds.has(id) };
  }

  const elapsed = performance.now() - startTime;
  return {
    algorithmName: 'Minimum Dominating Set (Strategic Relief Hub Placement)',
    executionTimeMs: Math.round(elapsed * 100) / 100,
    summary: `Selected ${dominatingHubIds.size} optimal emergency relief hubs / siren coverage centers to guarantee 100% population reachability within ${coverageRadiusKm} km.`,
    steps,
    nodeAttributes: nodeAttrs,
    customMetrics: {
      'Total Emergency Hubs': dominatingHubIds.size,
      'Total Covered Population': Object.values(graph.nodes).reduce((acc, n) => acc + n.population, 0).toLocaleString(),
      'Coverage Efficiency': `${Math.round((allNodeIds.length / dominatingHubIds.size) * 10) / 10} zones / hub`,
      'Population Coverage Rate': '100.0%'
    }
  };
}

File: backend/algorithms/dominating_set.py
import math
import time
from typing import Dict, List, Set, Any
from ..models.graph import GraphData, GraphNode
from ..models.algorithm_results import AlgorithmExecutionResult, AlgorithmStep

def get_distance(node_a: GraphNode, node_b: GraphNode) -> float:
    d_lat = (node_b.lat - node_a.lat) * 111.32
    d_lng = (node_b.lng - node_a.lng) * 111.32 * math.cos(math.radians(node_a.lat))
    return math.sqrt(d_lat * d_lat + d_lng * d_lng)

def run_dominating_set_hub_placement(
    graph: GraphData,
    coverage_radius_km: float = 3.5
) -> AlgorithmExecutionResult:
    start_time = time.perf_counter()
    steps: List[AlgorithmStep] = []
    all_node_ids = list(graph.nodes.keys())
    covered_nodes: Set[str] = set()
    dominating_hub_ids: Set[str] = set()

    # Precompute coverage neighborhood
    neighborhood: Dict[str, Set[str]] = {}
    for u in all_node_ids:
        neighborhood[u] = {u}
        node_u = graph.nodes[u]

        # Topological neighbors
        for edge_id in graph.adjacencyList.get(u, []):
            edge = graph.edges.get(edge_id)
            if not edge or edge.isBlocked:
                continue
            v = edge.target if edge.source == u else edge.source
            neighborhood[u].add(v)

        # Spatial radius neighbors
        for v in all_node_ids:
            if u == v:
                continue
            node_v = graph.nodes[v]
            if get_distance(node_u, node_v) <= coverage_radius_km:
                neighborhood[u].add(v)

    steps.append(AlgorithmStep(
        stepIndex=0,
        description=f"Dominating Set optimizer initialized. Target: Determine minimum strategic relief hub locations covering all {len(all_node_ids)} vertices within $\\le {coverage_radius_km}$ km radius.",
        highlightedNodeIds=[],
        highlightedEdgeIds=[]
    ))

    # Greedy set cover
    while len(covered_nodes) < len(all_node_ids):
        best_candidate = None
        max_newly_covered = -1

        for candidate in all_node_ids:
            node = graph.nodes[candidate]
            bonus = 2 if node.type in ('shelter', 'hospital', 'depot') else 0
            new_cov = sum(1 for nbr in neighborhood[candidate] if nbr not in covered_nodes)
            score = new_cov + bonus

            if score > max_newly_covered:
                max_newly_covered = score
                best_candidate = candidate

        if not best_candidate or max_newly_covered <= 0:
            for u in all_node_ids:
                if u not in covered_nodes:
                    dominating_hub_ids.add(u)
                    covered_nodes.add(u)
            break

        dominating_hub_ids.add(best_candidate)
        newly_covered_in_step = []
        for nbr in neighborhood[best_candidate]:
            if nbr not in covered_nodes:
                covered_nodes.add(nbr)
                newly_covered_in_step.append(nbr)

        if len(steps) < 30:
            steps.append(AlgorithmStep(
                stepIndex=len(steps),
                description=f"Selected '{graph.nodes[best_candidate].name}' as Strategic Hub #{len(dominating_hub_ids)}. Covered {len(newly_covered_in_step)} new neighborhood zones.",
                highlightedNodeIds=[best_candidate] + newly_covered_in_step[:4],
                highlightedEdgeIds=[]
            ))

    node_attrs: Dict[str, Dict[str, Any]] = {
        nid: {'isDominatingHub': (nid in dominating_hub_ids)} for nid in all_node_ids
    }
    elapsed = (time.perf_counter() - start_time) * 1000

    total_pop = sum(n.population for n in graph.nodes.values())

    return AlgorithmExecutionResult(
        algorithmName='Minimum Dominating Set (Strategic Relief Hub Placement - Python)',
        executionTimeMs=round(elapsed, 2),
        summary=f"Selected {len(dominating_hub_ids)} optimal emergency relief hubs / siren centers to guarantee 100% population coverage within {coverage_radius_km} km.",
        steps=steps,
        nodeAttributes=node_attrs,
        customMetrics={
            'Total Strategic Hubs': len(dominating_hub_ids),
            'Total Covered Population': f"{total_pop:,}",
            'Coverage Efficiency': f"{len(all_node_ids) / max(len(dominating_hub_ids), 1):.1f} zones/hub",
            'Population Coverage': '100.0%',
            'Execution Speed': f"{round(elapsed, 2)} ms"
        }
    )
